subsys/net/lib/lwm2m/lwm2m_shell.c

```diff
@@ -62,18 +86,367 @@ static int cmd_send(const struct shell *sh, size_t argc, char **argv)
 	}
 
 	ret = lwm2m_engine_send(ctx, (const char **)&(argv[ignore_cnt]),
-			path_cnt, confirmable);
+				path_cnt, confirmable);
 	if (ret < 0) {
 		shell_error(sh, "can't do send operation, request failed\n");
 		return -ENOEXEC;
 	}
+	return 0;
+}
+
+static int cmd_exec(const struct shell *sh, size_t argc, char **argv)
+{
+	struct lwm2m_ctx *ctx = lwm2m_rd_client_ctx();
+
+	if (!ctx) {
+		shell_error(sh, "no lwm2m context yet\n");
+		return -ENOEXEC;
+	}
+
+	int ignore_cnt = 2; /* Subcmd + PATH */
+	const char *pathstr = argv[1];
+	struct lwm2m_obj_path path;
+	int ret = lwm2m_string_to_path(pathstr, &path, '/'); /* translate path -> path_obj */
+
+	if (ret < 0) {
+		shell_error(sh, "Illegal path (PATH %s)\n", pathstr);
+		return -EINVAL;
+	}
+
+	struct lwm2m_engine_res *res = lwm2m_engine_get_res(&path);
+
+	if (ret < 0) {
+		shell_error(sh, "Resource not found (err %d)\n", ret);
+		return -EINVAL;
+	}
+
+	if (!res->execute_cb) {
+		shell_error(sh, "No execute callback\n!");
+		return -EINVAL;
+	}
+
+	ret = res->execute_cb(path.obj_inst_id, argv[ignore_cnt],
+			      argc - ignore_cnt);
+	if (ret < 0) {
+		shell_error(sh, "returned (err %d)\n", ret);
+		return -ENOEXEC;
+	}
+
+	return 0;
+}
+
+static int cmd_read(const struct shell *sh, size_t argc, char **argv)
+{
+	struct lwm2m_ctx *ctx = lwm2m_rd_client_ctx();
+
+	if (!ctx) {
+		shell_error(sh, "no lwm2m context yet\n");
+		return -ENOEXEC;
+	}
+
+	if (argc < 2) {
+		shell_error(sh, "no arguments or path(s)\n");
+		shell_help(sh);
+		return -EINVAL;
+	}
+	const char *dtype = "-s"; /* default */
+	const char *pathstr = argv[1];
+	int ret = 0;
+
+	if (argc > 2) { /* read + path + options(data type) */
+		dtype = argv[2];
+	}
+	if (strcmp(dtype, "-s") == 0) {
+		const char *buff;
+		uint16_t buff_len = 0;
+
+		ret = lwm2m_engine_get_res_buf(pathstr, (void **)&buff,
+					       &buff_len, NULL, NULL);
+		if (ret != 0) {
+			goto out;
+		}
+		ret = lwm2m_engine_set_res_data_len(pathstr, buff_len + 1);
+		if (ret != 0) {
+			goto out;
+		}
```

Comment:
```suggestion
```
No need to touch the buffer length at all.
We are not modifying here, just reading.